Month1-Week2-Section2

camodesigns opened the pull request.

camodesigns committed:
FirstCommin Section2 Week2

camodesigns committed:
Moth1-Week2-Section2 FoirstCommit

Firts Commit

camodesigns committed:
eliminatedall

camodesigns committed:
Month1-Week2-Section2 All the section

Spotify App Complete
Playlist of test in the  app added
Input test in all

camodesigns committed:
Month1-Week2-Section2 Update

Const Added,
cin.ignore() adeed for  a problem  cin.peek()

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Song.h:
```diff
@@ -0,0 +1,26 @@
+class FSong
+{
+private:
+    int MinutesDuration;
```
Please follow the exercise's instructions closely. You were supposed to just have one variable for the duration which holds the total seconds of the song. It is important to differentiate the way we want to store our data and the way we want to display it. Of course, to display the duration, the format `minutes:seconds`, but it is far less complicated to store the data as just one variable. That way there aren't different values storing the same data.

camodesigns replied:
Changed

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Song.h:
```diff
@@ -0,0 +1,26 @@
+
+    FSong(int EntryMinutesDuration, int EntrySecondsDuration, std::string EntryArtistName, std::string EntrySongName);
+    std::string GetFormatedDuration()const;
+    bool bIsvalid();
```
Remember to use PascalCase and functions don't need to have a b prefix. The documentation has some examples for boolean functions.
![image](https://user-images.githubusercontent.com/36243539/181571456-e3ead0f1-d5d3-49c5-a6f4-de85879a6b4c.png)

camodesigns replied:
ok, I understand, changed

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Song.h:
```diff
@@ -0,0 +1,26 @@
+#pragma once
+#include <string>
+#include <iostream>
+#include <string>
```
You're including string twice

camodesigns replied:
OMG, I Look, #include <string>Deleted

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Song.cpp:
```diff
@@ -0,0 +1,70 @@
+    ArtistName = "";
+    SongName = "";
+}
+FSong::FSong(int EntryMinutesDuration, int EntrySecondsDuration, std::string EntryArtistName, std::string EntrySongName)
```
Remember that parameters should be const if they are not modified inside the function. Please make this change for the whole review

camodesigns replied:
Ok, Understanded, and Added in all the program

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Song.cpp:
```diff
@@ -0,0 +1,70 @@
+
+}
+
+
```
Remember that consistency is important. Here you have 3 spaces between functions, then you have none, then you have one, and then you have none for the rest of the file.

I consider one space to be the most legible and clear.

camodesigns replied:
Changed

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Song.cpp:
```diff
@@ -0,0 +1,70 @@
+    FormatedDuration = "Minutes: " + std::to_string(MinutesDuration) + "\t" "Seconds" + std::to_string(SecondsDuration);
+    return FormatedDuration;
+}
+bool FSong::bIsvalid()
```
This method should be const, since it's not modifying the object in any way :)

camodesigns replied:
Changed

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Song.cpp:
```diff
@@ -0,0 +1,70 @@
+}
+bool FSong::bIsvalid()
+{
+    if (MinutesDuration <= 0 && SecondsDuration < 0)
```
This code works perfectly fine, but it is a bit complex for how trivial it is. Instead of writing
``` 
if (bCondition)
{
    return true;
}
else
{
    return false;
}
```
you could just write
```
return bCondition;
```
So, in this case, you could just write:
```
return MinutesDuration > 0 && SecondsDuration >= 0;
```

That way your function is clearer and easier to read, and you saved 7 lines for such a simple function.

camodesigns replied:
Wow, I didn't know that method, I really thank you

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/PlayList.h:
```diff
@@ -0,0 +1,20 @@
+#pragma once
+#include "Song.h"
+
+class FPlayList {
```
Remember to follow conventions. This curly brace is not following our standards.

camodesigns replied:
changed

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/PlayList.h:
```diff
@@ -0,0 +1,20 @@
+#pragma once
+#include "Song.h"
+
+class FPlayList {
```
Also, "playlist" is just one word hehe, it should be FPlaylist instead of FPlayList

camodesigns replied:
all the PlayList changed for Playlist

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/PlayList.h:
```diff
@@ -0,0 +1,20 @@
+public:
+	FPlayList();
+	FPlayList(std::string NewPlayListTitle);
+	std::string GetListTitle();
```
Careful with these kinds of abbreviations. `GetTitle()` and `SetTitle(...)` is more than enough since the context already makes it clear that you mean the playlist title. However, if you really want to be specific, writing `List` instead of `Playlist` reduces clarity and makes the name ambiguous, especially considering what I commented before saying that Playlist is just one word.

camodesigns replied:
Modified

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/PlayList.h:
```diff
@@ -0,0 +1,20 @@
+	void SetListTitle(std::string NewPlayListTitle);
+	bool bAddSong(FSong FSong);
+	FSong GetSong(int SongIndex);
+	void  PlayListDisplay();
```
Remember that functions should start with a verb since they should describe what they do. This name would be just fine if it was written the other way around. `DisplayPlaylist` is a lot clearer than `PlaylistDisplay`

camodesigns replied:
modified

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/PlayList.cpp:
```diff
@@ -0,0 +1,52 @@
+	std::cout << "PlayList" << std::endl;
+	std::cout << "Song Title\t\t\t\tArtist Name\t\t\t\tDuration" << std::endl;
+	std::cout << "------------------------------------------------------------------------------------------" << std::endl;
+	for (int CurrentSong = 0; CurrentSong < SongCount; CurrentSong++)
```
This is an example of misleading names. I know you like to name your indices explicitly and that's actually very nice. However, keep in mind that they are still _indices_. `CurrentSong` is not actually the current song, but rather the current song _index_. By itself, this can be misleading, but it becomes a bigger problem since now you have used that name, and so you can't use it for anything else.

For that reason, when you actually get the song using the index, you can't name it `CurrentSong` so you had to name it `ListedSong`. Now, by just reading the names, would you know what the difference between `CurrentSong` and `ListedSong` is? If it was me, I'd assume that `CurrentSong` is the actual current song and I would have no idea what `ListedSong` is, which is not good 😅 .

It's cases like this one that show why being explicit is important. `CurrentSong` should instead be `CurrentSongIndex` or something like that, and `ListedSong` should be `CurrentSong`. What do you think?

camodesigns replied:
It's reasonable, thank you, changed

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/PlayListContainer.h:
```diff
@@ -0,0 +1,16 @@
+	int PlayListCount;
+public:
+	FPlayListContainer();
+	FPlayListContainer(FPlayList NewPlayList);
```
Do you really need a constructor that takes in a playlist for a container of playlists?

camodesigns replied:
You havve reason, not necesary I delete that constructor

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/PlayListContainer.cpp:
```diff
@@ -0,0 +1,25 @@
+FPlayListContainer::FPlayListContainer(FPlayList NewPlayList)
+{
+	PlayLists[PlayListCount] = NewPlayList;
+	PlayListCount++;
```
Careful. PlaylistCount has not been initialized at this point, it could be any random value. PlaylistCount++ is not guaranteed to be 1 every time.

camodesigns replied:
I delete that constructor and I change  the PlaylistCount in FPlaylistContainer::AddPlaylist(FPlaylist NewPlaylist) for PlaylistCount +1

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Spotify.cpp:
```diff
@@ -0,0 +1,241 @@
+        }
+        ActiveCommand = (MenuCommand)OptionChoice;
+        system("CLS");
+    } while (!std::cin.good());
```
Doesn't seem like you need this `do-while` since you already have another while checking for this same condition.

camodesigns replied:
Changed

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Spotify.cpp:
```diff
@@ -0,0 +1,241 @@
+        std::cout << "Invalid input, please enter Your playlist name again: " << std::endl;
+        std::getline(std::cin, PlayListName);
+    }
+    FPlayList ANewPlayList(PlayListName);
```
Why do you need that `A` prefix? Why not `NewPlaylist`?

camodesigns replied:
changed

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Spotify.cpp:
```diff
@@ -0,0 +1,241 @@
+        std::cin.clear();
+        std::cin.ignore(1000, '\n');
+        system("CLS");
+    } while (AddSongs == 1);
```
AddSongs is not an acceptable name for a variable that represents the menu choice of the user. 👀

camodesigns replied:
Changed

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Spotify.cpp:
```diff
@@ -0,0 +1,241 @@
+    } while (AddSongs == 1);
+    PlayListContainer.AddPlaylist(ANewPlayList);
+    ActiveCommand = MenuCommand::MainMenu;
+}
```
This method is very nice! But I would suggest using a bit more spaces. It is hard to follow the function since there are hardly any spaces, so you get no breaks when reading through it.

camodesigns replied:
Spaces Added

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Spotify.cpp:
```diff
@@ -0,0 +1,241 @@
+        std::cout << "Well, now add a song to " << PlayListName << " Playlist" << std::endl;
+        std::cout << "Now,Follow the instructions" << std::endl;
+
+        int EntryMinutesDuration;
```
Do you really need to declare all these variables here? Why not declare them closer to where they are actually defined?

camodesigns replied:
reubicated

JDAlzate commented on Month1/Week2/Section2/Month1-Week2-Section2/Spotify/Spotify.cpp:
```diff
@@ -0,0 +1,241 @@
+    int Choice = 1;
+    do{
+        system("CLS");
+     for (int ListWalker = 0;ListWalker < 5; ListWalker++)
```
Again, it's nice that you want to make your indices more explicit, but it loses it's purpose if the name is not clearer than `i` 😅 . ListWalker doesn't really tell me anything. Instead, you should make it clear that this is an index, so what about `PlaylistIndex`?

camodesigns replied:
Changed

JDAlzate reviewed (changes requested):
Nice job! You're app is looking pretty good :) But there are a few things that do need to be changed.

camodesigns committed:
Month1-Week2-Section2-Updates

Method Changes
Names change
Add and delete spaces
change variables names
add and delete variables
